src/node_api.cc

```diff
@@ -2766,7 +2766,26 @@ class Work {
     Work* work = static_cast<Work*>(req->data);
 
     if (work->_complete != nullptr) {
-      work->_complete(work->_env, ConvertUVErrorCode(status), work->_data);
+      napi_env env = work->_env;
+
+      // Establish a handle scope here so that every callback doesn't have to.
+      // Also it is needed for the exception-handling below.
+      v8::HandleScope scope(env->isolate);
+
+      work->_complete(env, ConvertUVErrorCode(status), work->_data);
```

Comment:
So … I’ve given this a bit of thought before. Semantically and technically it would make the most sense if `_complete` ran inside a `MakeCallback` scope. Unfortunately, right now that would mean that `_complete` would need to be specified as a JS function (which would probably even be fine except that nobody would remember to call `napi_delete_async_work` inside it).

Once #11883 lands, I would like to look into splitting the `enter` and `exit` parts of `MakeCallback` into their own parts (like what https://github.com/matthewloring/node/commit/dd178336a60c2698619b96a62267bd528c09ae0d does for Promises), and ideally using them here to give the users an easier API.

Do you think that’s a good idea? I think it makes sense but it might require some weird hacks in N-API for the time being (like wrapping the `_complete` callback inside a JS function just to pass it to `MakeCallback`…)